serviceregistration/consul/consul_service_registration.go

```diff
@@ -212,10 +211,21 @@ func NewConsulServiceRegistration(conf map[string]string, logger log.Logger) (sr
 		checkTimeout:        checkTimeout,
 		disableRegistration: disableRegistration,
 
-		notifyActiveCh:      make(chan notifyEvent),
-		notifySealedCh:      make(chan notifyEvent),
-		notifyPerfStandbyCh: make(chan notifyEvent),
+		notifyActiveCh:      make(chan bool),
+		notifySealedCh:      make(chan bool),
+		notifyPerfStandbyCh: make(chan bool),
+
+		isActive:      state.IsActive,
+		isSealed:      state.IsSealed,
+		isPerfStandby: state.IsPerformanceStandby,
 	}
+	go func() {
+		if err := c.runServiceRegistration(&sync.WaitGroup{}, shutdownCh, redirectAddr); err != nil {
+			if logger.IsError() {
+				logger.Error(fmt.Sprintf("error running service registration: %s", err))
+			}
+		}
+	}()
```

Comment:
I generally prefer it when async APIs have a `Run` & `Stop` method. Which allows you to control the execution outside of the context of the constructor. I also like it when `Run` doesn't spawn a go routine for you and instead leaves that up to the caller. What do you think about adopting that strategy?

Reply:
I'm flexible about it. Need some clarification, though.

What would be the advantage to having a `Run` method? Would it be in addition to the `Factory` method (which is basically a constructor), or would it be in lieu of it?

An when you say a `Stop` method, would it basically be an `OnStop` method? If we added it, does that mean we wouldn't pass in the `shutdownCh` because instead they'd use `Stop`?

Reply:
Also, might be worth noting that the observer pattern doesn't tend to use a `Run` method. Here are a couple of examples:

- https://gist.github.com/patrickmn/1549985
- https://www.baeldung.com/java-observer-pattern
- https://refactoring.guru/design-patterns/observer

If you're thinking of a different pattern, it would be interesting to hear about the principle and your reasoning.

Reply:
Run method was added here: https://github.com/hashicorp/vault/pull/8123/commits/7070a6a8451bcfa37387e57a2f2f9d39ee5bc34d